File: src/main/java/net/hotamachisubaru/digger/sqlite/SQLiteDatabase.java
package net.hotamachisubaru.digger.sqlite;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import net.hotamachisubaru.digger.Digger;

import java.sql.*;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class SQLiteDatabase{
    private Connection connection;

    public void openConnection(String path) throws SQLException {
        // データベース接続の確立
        String url = "jdbc:sqlite:" + path + "/Database.db";
        connection = DriverManager.getConnection(url);

        // テーブルの作成
        try (Statement statement = connection.createStatement()) {
            String playerDataTableCreationQuery = "CREATE TABLE IF NOT EXISTS player_data ("
                    + "UUID VARCHAR(255) NOT NULL,"
                    + "PlayerName VARCHAR(255),"
                    + "BlocksMined INT DEFAULT 0,"
                    + "PRIMARY KEY (UUID));";
            statement.execute(playerDataTableCreationQuery);

            String placedBlocksTableCreationQuery = "CREATE TABLE IF NOT EXISTS placed_blocks ("
                    + "UUID INTEGER PRIMARY KEY AUTOINCREMENT,"
                    + "World VARCHAR(255),"
                    + "X INT,"
                    + "Y INT,"
                    + "Z INT);";
            statement.execute(placedBlocksTableCreationQuery);
        }
    }

    // データベースからデータを取得するメソッド
    public static Map<UUID, Digger.PlayerData> getData() throws SQLException {
        Map<UUID, Digger.PlayerData> data = new HashMap<>();
        String query = "SELECT * FROM player_data"; // データベースのテーブル名を適宜変更してください

        try (PreparedStatement stmt = connection.prepareStatement(query);
             ResultSet rs = stmt.executeQuery()) {

            while (rs.next()) {
                // データベースからUUIDとブロックの採掘数を取得
                String uuidString = rs.getString("UUID");
                UUID uuid = UUID.fromString(uuidString);
                String playerName = rs.getString("PlayerName");
                int blocksMined = rs.getInt("BlocksMined");

                // 取得したデータでPlayerDataオブジェクトを作成
                Digger.PlayerData playerData = new Digger.PlayerData(playerName, blocksMined);
                data.put(uuid, playerData);
            }
        }

        return data;
    }

    public Map<UUID, Digger.PlayerData> loadData() throws SQLException {
        Map<UUID, Digger.PlayerData> loadedData = new HashMap<>();
        // SQLクエリを実行し、結果を取得
        String query = "SELECT * FROM player_data;";
        try (Statement stmt = this.connection.createStatement();
             ResultSet results = stmt.executeQuery(query)) {

            while (results.next()) {
                UUID uuid = UUID.fromString(results.getString("UUID"));
                String playerName = results.getString("PlayerName");
                int blocksMined = results.getInt("BlocksMined");

                loadedData.put(uuid, new Digger.PlayerData(playerName, blocksMined));
            }
        }
        return loadedData;
    }

    public static void saveData(Map<UUID, Digger.PlayerData> blockCount, Iterable<Location> placedBlocks) throws SQLException {
        // blockCount の保存
        saveBlockCount(blockCount);


        // placedBlocks の保存
        savePlacedBlocks(placedBlocks);
    }

    private void saveBlockCount(Map<UUID, Digger.PlayerData> blockCount) throws SQLException {
        for (Map.Entry<UUID, Digger.PlayerData> entry : blockCount.entrySet()) {
            Bukkit.getLogger().info("UUIDのPlayerDataをセーブしています: " + entry.getKey() + ", PlayerName: " + entry.getValue().getPlayerName());
        }
        String blockCountQuery = "INSERT INTO player_data (UUID, BlocksMined, PlayerName) VALUES (?, ?, ?) "
                + "ON CONFLICT(UUID) DO UPDATE SET BlocksMined = excluded.BlocksMined, PlayerName = excluded.PlayerName;";

        try (PreparedStatement pstmt = connection.prepareStatement(blockCountQuery)) {
            for (Map.Entry<UUID, Digger.PlayerData> entry : blockCount.entrySet()) {
                pstmt.setString(1, entry.getKey().toString());
                pstmt.setInt(2, entry.getValue().getBlocksMined());
                pstmt.setString(3, entry.getValue().getPlayerName()); // PlayerNameの設定
                pstmt.executeUpdate();
            }
        }
    }


    private void savePlacedBlocks(Iterable<Location> placedBlocks) throws SQLException {
        String placedBlocksQuery = "INSERT INTO placed_blocks (World, X, Y, Z) VALUES (?, ?, ?, ?);";

        try (PreparedStatement pstmt = connection.prepareStatement(placedBlocksQuery)) {
            for (Location loc : placedBlocks) {
                pstmt.setString(1, loc.getWorld().getName());
                pstmt.setInt(2, loc.getBlockX());
                pstmt.setInt(3, loc.getBlockY());
                pstmt.setInt(4, loc.getBlockZ());
                pstmt.executeUpdate();
            }
        }
    }
    public static boolean checkConnection() throws SQLException {
        // connection はデータベースへの接続オブジェクトです
        return connection != null && !connection.isClosed();
    }

    public void closeConnection() throws SQLException {
        if (connection != null && !connection.isClosed()) {
            connection.close();
        }
    }
}

File: src/main/java/net/hotamachisubaru/digger/sqlite/SQLiteDatabase.kt
package net.hotamachisubaru.digger.sqlite

import net.hotamachisubaru.digger.Digger.PlayerData
import org.bukkit.Location
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException
import java.util.*

class SQLiteDatabase {
    private var connection: Connection? = null

    // データベース接続を開くメソッド
    @Throws(SQLException::class)
    fun openConnection(path: String) {
        // SQLiteデータベースに接続
        val url = "jdbc:sqlite:$path/Database.db"
        connection = DriverManager.getConnection(url)

        // テーブルの作成
        createTables()
    }

    // テーブルの作成
    @Throws(SQLException::class)
    private fun createTables() {
        connection!!.createStatement().use { statement ->
            // player_data テーブルの作成
            val playerDataTableCreationQuery = ("CREATE TABLE IF NOT EXISTS player_data ("
                    + "UUID VARCHAR(255) NOT NULL,"
                    + "PlayerName VARCHAR(255),"
                    + "BlocksMined INT DEFAULT 0,"
                    + "PRIMARY KEY (UUID));")
            statement.execute(playerDataTableCreationQuery)

            // placed_blocks テーブルの作成
            val placedBlocksTableCreationQuery = ("CREATE TABLE IF NOT EXISTS placed_blocks ("
                    + "ID INTEGER PRIMARY KEY AUTOINCREMENT,"
                    + "World VARCHAR(255),"
                    + "X INT,"
                    + "Y INT,"
                    + "Z INT);")
            statement.execute(placedBlocksTableCreationQuery)
        }
    }

    @get:Throws(SQLException::class)
    val data: Map<UUID, PlayerData>
        // データベースからデータを取得するメソッド
        get() {
            if (!checkConnection()) {
                throw SQLException("データベース接続が確立されていません。")
            }

            val data: MutableMap<UUID, PlayerData> = HashMap()
            val query = "SELECT * FROM player_data;"

            connection!!.prepareStatement(query).use { stmt ->
                stmt.executeQuery().use { rs ->
                    while (rs.next()) {
                        // データベースからUUIDとブロックの採掘数を取得
                        val uuid = UUID.fromString(rs.getString("UUID"))
                        val playerName = rs.getString("PlayerName")
                        val diamondMined = rs.getInt("DiamondMined")

                        // 取得したデータでPlayerDataオブジェクトを作成
                        val playerData = PlayerData(playerName, diamondMined)
                        data[uuid] = playerData
                    }
                }
            }
            return data
        }

    // データをロードするメソッド
    @Throws(SQLException::class)
    fun loadData(): Map<UUID, PlayerData> {
        return data
    }

    // データを保存するメソッド
    @Throws(SQLException::class)
    fun saveData(blockCount: Map<UUID, PlayerData>, placedBlocks: Iterable<Location>) {
        if (!checkConnection()) {
            throw SQLException("データベース接続が確立されていません。")
        }

        saveBlockCount(blockCount)
        savePlacedBlocks(placedBlocks)
    }

    // ブロックカウントデータを保存
    @Throws(SQLException::class)
    private fun saveBlockCount(blockCount: Map<UUID, PlayerData>) {
        val blockCountQuery = ("INSERT INTO player_data (UUID, BlocksMined, PlayerName) VALUES (?, ?, ?) "
                + "ON CONFLICT(UUID) DO UPDATE SET BlocksMined = excluded.BlocksMined, PlayerName = excluded.PlayerName;")

        connection!!.prepareStatement(blockCountQuery).use { pstmt ->
            for ((key, value) in blockCount) {
                pstmt.setString(1, key.toString())
                pstmt.setInt(2, value.blocksMined)
                pstmt.setString(3, value.playerName)
                pstmt.executeUpdate()
            }
        }
    }

    // 配置されたブロックデータを保存
    @Throws(SQLException::class)
    private fun savePlacedBlocks(placedBlocks: Iterable<Location>) {
        val placedBlocksQuery = "INSERT INTO placed_blocks (World, X, Y, Z) VALUES (?, ?, ?, ?);"

        connection!!.prepareStatement(placedBlocksQuery).use { pstmt ->
            for (loc in placedBlocks) {
                pstmt.setString(1, loc.world.name)
                pstmt.setInt(2, loc.blockX)
                pstmt.setInt(3, loc.blockY)
                pstmt.setInt(4, loc.blockZ)
                pstmt.executeUpdate()
            }
        }
    }

    // データベース接続がアクティブかどうかをチェック
    @Throws(SQLException::class)
    fun checkConnection(): Boolean {
        return connection != null && !connection!!.isClosed
    }

    // データベース接続を閉じるメソッド
    @Throws(SQLException::class)
    fun closeConnection() {
        if (connection != null && !connection!!.isClosed) {
            connection!!.close()
        }
    }

    companion object {
        // シングルトンインスタンスの取得メソッド
        var instance: SQLiteDatabase? = null
            get() {
                if (field == null) {
                    field = SQLiteDatabase()
                }
                return field
            }
            private set
    }
}
